### PR TITLE
Optimize `BigDecimal#div` for inexact divisions

### PR DESCRIPTION
**Pull request by HertzDevil**
`BigDecimal#div` is very slow when an inexact division is performed, because it allocates O(n) temporary `BigInt`s in the following loop:

https://github.com/crystal-lang/crystal/blob/2a5404f910a5c641cd4571acce67aa022d1d37cd/src/big/big_decimal.cr#L225-L230

This PR uses [`mpz_remove`](https://gmplib.org/manual/Number-Theoretic-Functions) to partially factorize the mantissas so that only O(1) `BigInt`s are needed to obtain the returned value's scale. This GMP function is wrapped by `BigInt#factor_by`, which is currently hidden, but depends on another helper method `#divisible_by?`, which in turn simply overrides the definition from `Int` (the base definition would allocate another `BigInt` through a call to `#remainder`). Consequently, `max_div_iterations` no longer refers to any iteration, but I have kept the same parameter name for now because we might not be able to change it for 1.x.

The performance improvements are immediately clear when inexact divisions are involved:

```crystal
require "big"
require "benchmark"

N = ENV["N"].to_i

struct BigDecimal
  # implementation before this PR
  def div_old(other, max_div_iterations); ...; end
end

macro test(method, prec)
  (N // 100).times do
    x = BigDecimal.new(Random.rand(1..10000), 3)
    y = BigDecimal.new(Random.rand(1..10000), 1) # 99.75% of them produce inexact divisions
    100.times do
      x.{{ method.id }}(y, {{ prec }})
    end
  end
end

Benchmark.ips do |b|
  b.report("div new (prec=9)") { test :div, 9 }
  b.report("div old (prec=9)") { test :div_old, 9 }
  b.report("div new (prec=15)") { test :div, 15 }
  b.report("div old (prec=15)") { test :div_old, 15 }
  b.report("div new (default)") { test :div, BigDecimal::DEFAULT_MAX_DIV_ITERATIONS }
  b.report("div old (default)") { test :div_old, BigDecimal::DEFAULT_MAX_DIV_ITERATIONS }
end
```

```
$ N=100000 bin/crystal run --release test.cr
 div new (prec=9)  22.86  ( 43.74ms) (± 5.75%)  14.7MB/op        fastest
 div old (prec=9)   4.51  (221.53ms) (± 7.46%)   116MB/op   5.06× slower
div new (prec=15)  22.78  ( 43.89ms) (± 7.05%)  14.7MB/op   1.00× slower
div old (prec=15)   2.83  (353.76ms) (± 4.45%)   190MB/op   8.09× slower
div new (default)  18.18  ( 55.01ms) (± 8.80%)  25.4MB/op   1.26× slower
div old (default) 425.79m (  2.35s ) (± 0.82%)  1.44GB/op  53.69× slower
```

Exact division occurs only when the divisor is a power of 2 multiplied by a power of 5. The first two commits alone would make division of small powers of 5 _slower_:

```crystal
macro test(method, prec)
  # ...
  x = BigDecimal.new(Random.rand(1..100), 1)
  y = BigDecimal.new(5, 0)
  # ...
end
```

```
$ N=500000 bin/crystal run --release test.cr
 div new (prec=9)   4.60  (217.19ms) (± 4.53%)  45.9MB/op   1.42× slower
 div old (prec=9)   6.41  (156.04ms) (± 5.78%)  63.7MB/op   1.02× slower
div new (prec=15)   4.50  (222.17ms) (± 6.78%)  45.9MB/op   1.45× slower
div old (prec=15)   5.87  (170.48ms) (± 6.35%)  63.6MB/op   1.11× slower
div new (default)   4.24  (235.97ms) (± 5.58%)  46.0MB/op   1.54× slower
div old (default)   6.52  (153.33ms) (± 6.73%)  63.6MB/op        fastest
```

Heuristics that improve this situation are contained in the third commit, by avoiding some calls to `factor_by(FIVE)`:

```
$ N=500000 bin/crystal run --release test.cr
 div new (prec=9)   6.83  (146.41ms) (± 7.83%)  39.9MB/op        fastest
 div old (prec=9)   6.52  (153.48ms) (± 8.78%)  63.5MB/op   1.05× slower
div new (prec=15)   6.75  (148.16ms) (± 8.25%)  39.9MB/op   1.01× slower
div old (prec=15)   6.59  (151.80ms) (± 6.28%)  63.6MB/op   1.04× slower
div new (default)   6.75  (148.05ms) (± 8.63%)  39.8MB/op   1.01× slower
div old (default)   6.64  (150.69ms) (± 8.54%)  63.6MB/op   1.03× slower
```

Another place that uses `factor_by` is the internal method `BigDecimal#factor_powers_of_ten`. It is used only in two places, but I did it anyway since it has the same issue as `#div`. ([`mpz_remove` itself](https://github.com/sethtroisi/libgmp/blob/master/mpz/remove.c) I believe has O((log n)^2) complexity.)

Related: #10641

### CODE REVIEW

File: src/big/big_decimal.cr [hunk 1]
```diff
@@ -205,31 +207,77 @@ struct BigDecimal < Number
   # ```
   def div(other : BigDecimal, max_div_iterations = DEFAULT_MAX_DIV_ITERATIONS) : BigDecimal
```
Comment (straight-shoota) [inline src/big/big_decimal.cr:208]: We should probably have a deprecated overload with named argument ˋmax_div_iterationsˋ and change the positional argument to ˋprecisionˋ. That should be a safe change.

File: src/big/big_decimal.cr [hunk 2]
```diff
@@ -15,8 +15,10 @@ end
 # The general idea and some of the arithmetic algorithms were adapted from
 # the MIT/APACHE-licensed [bigdecimal-rs](https://github.com/akubera/bigdecimal-rs).
 struct BigDecimal < Number
-  ZERO                       = BigInt.new(0)
-  TEN                        = BigInt.new(10)
+  private TWO  = BigInt.new(2)
+  private FIVE = BigInt.new(5)
+  private TEN  = BigInt.new(10)
```
Comment (straight-shoota) [inline src/big/big_decimal.cr:20]: Removing these constants from the public API is abreaking change. If we want to make them private, they can only be deprecated for now.
I'm not convinced we should hide them, though.

Comment (HertzDevil) [reply src/big/big_decimal.cr:20]: Deprecation it is, then. I don't think anyone really expects `BigDecimal`'s numeric constants not to be `BigDecimal`s though.

Comment (straight-shoota) [reply src/big/big_decimal.cr:20]: Yeah, that seems certainly correct 😆

**Review by straight-shoota — APPROVED** [review]

File: src/big/big_decimal.cr (hunk 2 above)
Comment (HertzDevil) [reply src/big/big_decimal.cr]: This part is actually resolved by #11820

**Pull request merged**